### PR TITLE
(286) Add Pact for contract testing

### PR DESCRIPTION
**Pull request by Gweaton**
## Context

https://trello.com/c/c812hQ09/286-set-up-pact-on-the-ui-side

## Changes in this PR

This adds [Pact](https://pact.io/) for testing the contracts between the API and this client. This allows us to catch any changes to the API that might break our service in our tests.

Running the `courseClient` test generates a Pact contract (stored locally in `pact/pacts`) to specify the shape of the response we expect from the API. We publish this to the [Pact Broker](https://pact-broker-prod.apps.live-1.cloud-platform.service.justice.gov.uk/) on successful CI builds, and tag them with the environment once deployed.

We'll need to configure the API to write tests to fulfil this contract and publish the results to the Pact broker once this is merged.

The published pact broker for our service is accessible [here](https://pact-broker-prod.apps.live-1.cloud-platform.service.justice.gov.uk/pacts/provider/Accredited%20Programmes%20API/consumer/Accredited%20Programmes%20UI/latest).

### CODE REVIEW

File: doc/adr/0002-use-pact-for-contract-testing.md [hunk 1]
```diff
@@ -0,0 +1,59 @@
...
+We're developing a service with a frontend client (the UI) that communicates
+with a separate API, written by different developers often working in parallel.
+When running integration tests, the UI team do this against a mocked version of
+the API which may not always be in the state we expect.
+
+We need to ensure that any changes to the API don't break existing
+functionality in the user interface. Although we could use Cypress to test the
+deployed application, this would have a number of drawbacks:
+
+- We have to maintain integration and e2e tests, which are functionally
```
Comment (yndajas) [inline doc/adr/0002-use-pact-for-contract-testing.md:20]: Is it worth spelling out `e2e`?

File: doc/adr/0002-use-pact-for-contract-testing.md [hunk 2]
```diff
@@ -0,0 +1,59 @@
...
+  similar, but different enough for there to be differences in implementation
+- We have to wait for a new version of the code to be deployed before running
+  our e2e tests, meaning the feedback loop for noticing a problem and fixing it
+  is long
+- e2e tests are slow running, slowing down our build pipeline
+- As the tests run in a real environment, we are filling the database with test
+  data, which has to be cleaned out
+- The tests can be flaky, as pass for reasons that are outside our control,
```
Comment (yndajas) [inline doc/adr/0002-use-pact-for-contract-testing.md:28]: `and pass`?

File: doc/adr/0002-use-pact-for-contract-testing.md [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+  such as upstream APIs being down, or performance issues
+- When tests do fail, it's not always clear _what_ has broken, and it can be
+  hard to track down the source of the failure
+
+## Decision
+
+With all this in mind, we've decided to use [Pact](https://pact.io/) to carry
+out Contract Testing when testing our API client classes.
```
Comment (yndajas) [inline doc/adr/0002-use-pact-for-contract-testing.md:36]: Is there a reason for capitalising `Contract Testing`?

File: doc/adr/0002-use-pact-for-contract-testing.md [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+CI build. "The provider" then pulls these contracts and verifies its own
+response shape against them to ensure it's sending what's expected to the UI.
+If the shape is different to what the consumer expects, the build will fail and
+prevent us merging anything that could break functionality.
+
+## Consequences
+
+This will hopefully speed up our development, and raise any deviation in the
+contract between the API and the UI application in a timely manner without the
+need for expensive and flaky E2E tests. When API changes do cause the tests to
```
Comment (yndajas) [inline doc/adr/0002-use-pact-for-contract-testing.md:49]: `e2e` is now `E2E` - I think it's worth spelling out in at least the first instance, and then either using a provided abbreviation or continuing to spell it out

**Review by yndajas — APPROVED** [review]
A few small suggestions but in general it looks good to me

File: doc/adr/0002-use-pact-for-contract-testing.md (hunk 3 above)
Comment (Gweaton) [reply doc/adr/0002-use-pact-for-contract-testing.md]: I thought it was a name like "Test Driven Development" but I don't think it is, actually!

File: doc/adr/0002-use-pact-for-contract-testing.md (hunk 4 above)
Comment (Gweaton) [reply doc/adr/0002-use-pact-for-contract-testing.md:49]: Yeah I've spelt this out above and made it consistent throughout.

**Pull request merged**